### PR TITLE
Authentication APIs

### PR DESCRIPTION
**Pull request by HiAkshat**

### CODE REVIEW

**Review by pratik-edu — APPROVED** [review]

File: src/dao/demo.dao.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
-import demoModel from '@/models/demo.model';
+// import demoModel from '@/models/user.model';
 
 class DemoDao {
```
Comment (pratik-edu) [inline src/dao/demo.dao.ts:3]: Remove This File, No Longer Required !!

File: src/helpers/common.helper.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+export default class CommonHelper {}
```
Comment (pratik-edu) [inline src/helpers/common.helper.ts:1]: Needed??

File: src/config/index.ts [hunk 3]
```diff
@@ -34,6 +28,9 @@ export const services = (() => {
   return nconf.get('services') as IServices;
 })();
 
+export const email = (() => {
+  return nconf.get('email') as IEmailConfig;
+})();
```
Comment (HiAkshat) [inline src/config/index.ts:33]: add space below

File: src/services/authenticate.service.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+   * used to identify the user.
+   * @param {string} password - A string representing the user's password that needs to be validated.
+   * @returns An object containing the session token, expiration date, and user details
+   * (excluding the password), or an error message if authentication fails.
+   */
+  public async userLogin(identifier: string, password: string) {
+    // Find user based on identifier (phone_number, email, user_id)
+    const user = await this.userDAO.findByIdentifier(getUserIdentifierType(identifier), identifier);
+
+    if (!user || !(await comparePasswords(password, user.password))) {
```
Comment (HiAkshat) [inline src/services/authenticate.service.ts:59]: does `user.password `need to be decrypted before comparing or is this fine?

File: src/utils/auth.utils.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import bcrypt from 'bcrypt';
+import { v4 as uuidv4 } from 'uuid';
+
+/**
+ * Hashes a password using bcrypt.
+ * @param password - The password to hash.
+ * @returns The hashed password.
+ */
+export const hashPassword = async (password: string): Promise<string> => {
+  return await bcrypt.hash(password, 10);
```
Comment (pratik-edu) [inline src/utils/auth.utils.ts:10]: move this cost factor to config.
or it is better that cost factor should be random always.
whats your opinion

File: src/services/authenticate.service.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+    // Check for an existing valid OTP
+    const existingOtp = await this.otpDAO.findValidOtp(user._id);
+    if (existingOtp) {
+      await emailHelper.sendOtpEmail(user.email, existingOtp.otp);
+      return { message: 'OTP resent to your email', otp: existingOtp.otp };
+    }
+
+    // Generate a new OTP if no valid OTP exists
+    const otp = Math.floor(100000 + Math.random() * 900000).toString();
```
Comment (HiAkshat) [inline src/services/authenticate.service.ts:99]: In utils, add this and call to generate OTP.
```suggestion
    const getRandomNumber = (length = 6)  => {
    return Math.floor(Math.pow(10, length-1) + Math.random() * (Math.pow(10, length) - Math.pow(10, length-1) - 1));
}
```

File: src/services/authenticate.service.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    // Save the new OTP in the database
+    await this.otpDAO.createOtp({
+      user: user._id,
+      user_id: user.user_id,
+      email: user.email,
+      otp,
+      expiresAt,
+    });
+
+    await emailHelper.sendOtpEmail(user.email, otp);
```
Comment (HiAkshat) [inline src/services/authenticate.service.ts:112]: create private variable for Email helper. Don't export instance.

File: src/dao/otp.dao.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import OtpModel from '@/models/otp.model';
+import { IOtpSchema } from '@/interfaces/user.interface';
+
+class OtpDAO {
```
Comment (HiAkshat) [inline src/dao/otp.dao.ts:4]: create private variable for otp model

File: src/dao/session.dao.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import { SessionModel } from '@/models/sessions.model';
+
+export class SessionDAO {
```
Comment (HiAkshat) [inline src/dao/session.dao.ts:3]: create private variable for session model

File: src/middlewares/session.middleware.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import AuthenticateService from '@/services/authenticate.service';
```
Comment (HiAkshat) [inline src/middlewares/session.middleware.ts:1]: Add import comments

File: src/middlewares/validator.middleware.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
+import { NextFunction, Request, Response } from 'express';
```
Comment (HiAkshat) [inline src/middlewares/validator.middleware.ts:1]: Add import comments

File: src/controllers/authenticate.controller.ts [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    const { identifier, password } = req.body;
+
+    const loginResult = await this.authenticateService.userLogin(identifier, password);
+
+    const { sessionToken, expiresAt, user } = loginResult;
+    const maxAge = expiresAt ? expiresAt.getTime() - Date.now() : undefined;
+
+    //SETTING COOKIES AND HEADERS REQUIRED FOR OTHER APIS (FE SHOULD  NESSARY SEND THIS AS HEADERS IN EVERY API CALLS)
+    res.cookie('session_token', sessionToken, { maxAge });
+    res.setHeader('wm_usertype', user?.user_type || 'user');
```
Comment (HiAkshat) [inline src/controllers/authenticate.controller.ts:66]: we should probably replace wm with something else

Comment (pratikkhulge-blitz) [reply src/controllers/authenticate.controller.ts:66]: recommend!! wm seems little professional

**Review by pratik-edu — APPROVED** [review]

**Pull request merged**